xdgopenproxy/portal_launcher.go

```diff
@@ -0,0 +1,174 @@
+	err := bus.BusObject().Call("org.freedesktop.DBus.StartServiceByName", 0, desktopPortalBusName, uint32(0)).Store(&startResult)
+	if dbusErr, ok := err.(dbus.Error); ok {
+		// If it is not possible to activate the service
+		// (i.e. there is no .service file or the systemd unit
+		// has been masked), assume it is already
+		// running. Subsequent method calls will fail if this
+		// assumption is false.
+		if dbusErr.Name == "org.freedesktop.DBus.Error.ServiceUnknown" || dbusErr.Name == "org.freedesktop.systemd1.Masked" {
+			err = nil
+			startResult = 2
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+	switch startResult {
+	case 1: // DBUS_START_REPLY_SUCCESS
+	case 2: // DBUS_START_REPLY_ALREADY_RUNNING
+	default:
+		return nil, fmt.Errorf("unexpected response from StartServiceByName (code %v)", startResult)
+	}
+	return bus.Object(desktopPortalBusName, desktopPortalObjectPath), nil
+}
+
+// portalResponseSuccess is a numeric value indicating a success carrying out
+// the request, returned by the `response` member of
+// org.freedesktop.portal.Request.Response signal
+const portalResponseSuccess = 0
+
+// timeout for asking the user to make a choice, same value as in usersession/userd/launcher.go
+var defaultPortalRequestTimeout = 5 * time.Minute
+
+func (p *portalLauncher) portalCall(bus *dbus.Conn, call func() (dbus.ObjectPath, error)) error {
+	// see https://flatpak.github.io/xdg-desktop-portal/portal-docs.html for
+	// details of the interaction, in short:
+	// 1. caller issues a request to the desktop portal
+	// 2. desktop portal responds with a handle to a dbus object capturing the Request
+	// 3. caller waits for the org.freedesktop.portal.Request.Response
+	// 3a. caller can terminate the request earlier by calling close
+
+	// set up signal handling before we call the portal, so that we do not
+	// miss the signals
+	signals := make(chan *dbus.Signal, 1)
+	bus.Signal(signals)
+	defer func() {
+		bus.RemoveSignal(signals)
+		close(signals)
+	}()
+
+	// TODO: this should use dbus.Conn.AddMatchSignal, but that
+	// does not exist in the external copies of godbus on some
+	// supported platforms.
+	const matchRule = "type='signal',sender='" + desktopPortalBusName + "',interface='" + desktopPortalRequestIface + "',member='Response'"
+	if err := bus.BusObject().Call("org.freedesktop.DBus.AddMatch", 0, matchRule).Store(); err != nil {
+		return err
+	}
+	defer bus.BusObject().Call("org.freedesktop.DBus.RemoveMatch", 0, matchRule)
+
+	requestPath, err := call()
+	if err != nil {
+		return err
+	}
+	request := bus.Object(desktopPortalBusName, requestPath)
+
+	timeout := time.NewTimer(defaultPortalRequestTimeout)
+	defer timeout.Stop()
+	for {
+		select {
+		case <-timeout.C:
+			request.Call(desktopPortalRequestIface+".Close", 0).Store()
+			return &responseError{msg: "timeout waiting for user response"}
+		case signal := <-signals:
+			if signal.Path != requestPath || signal.Name != desktopPortalRequestIface+".Response" {
```

Comment:
I hadn't created separate consts for individual method names because they were short and didn't see much reuse.  I do see that this particular one is repeated three times between the code and tests though.

Do you think it is worth adding named constants for all of them?

Reply:
Nah, I'm just nitpicking. It's ok as it is.